### PR TITLE
Panels loading on scroll

### PR DESCRIPTION
**Pull request by jifwin**
This implements panels loading on scroll.

### CODE REVIEW

**Comment by jifwin** [comment]
Sorry for messing up CLA, I've accidentally used two accounts to commit. It should be fine now.

**Comment by jifwin** [comment]
Updated the PR to make it configurable.

**Comment by daniellee** [comment]
@jifwin this looks very good so far. The big decision is whether it should be configurable and if it should be on or off per default. Just going to test a bit more and I should be able to merge it tomorrow if no changes are needed.

**Comment by daniellee** [comment]
@jifwin we (Torkel and I) made some changes to your PR before merging. We decided that this is the default behavior so we removed the toggle. We moved the logic up a level to the panel_ctrl so that it works for all panels. 

It was a very good catch by you to notice that this would affect png rendering (dashboard-solo). That would have been easy to miss.

We'll see if we get any feedback during the beta phase of the release 4.3.0 about if the toggle is needed or not. (and if we missed any edge cases)

Thank you very much for the PR!

**Review by daniellee — COMMENTED** [review]
Great feature. As there were a few difficult decisions to make (should there be a toggle and whether it should apply to metric panels or all panels), we made a few changes ourselves rather than just giving feedback.

File: public/app/features/panel/panel_directive.ts [hunk 1]
```diff
@@ -57,7 +57,7 @@ var panelTemplate = `
   </div>
 `;
 
-module.directive('grafanaPanel', function($rootScope) {
+module.directive('grafanaPanel', function($rootScope, $document, $timeout) {
```
Comment (jifwin) [inline public/app/features/panel/panel_directive.ts]: Yes. I used to delay the refresh call, but then replaced it with debounce and forgot to remove timeout.

**Comment by torkelo** [comment]
removed the debounce, it caused panels not to load until you stopped scrolling and gave a very slow impression. Was there a reason for it?

**Comment by jifwin** [comment]
It was intended. Let's say I have a graph at the bottom of the dashboard. I need to scroll down to it and I don't want all the other graphs to load, since I don't need them. That's why I added a debounce. Also it reduces load on the data source. To make it faster we could set the debounce to let's say 100 ms. I know you decided to make load on scroll a default behaviour (non-configurable), but how about making the debounce configurable? I think a lot of users would expect that.

**Comment by torkelo** [comment]
I think that case (scroll down to bottom) is very edge case so not so important. 

The problem is that even a small debounce causes a really bad experience as no graphs will load if you continuously scroll slowly down.

**Comment by tmlippman** [comment]
I'd appreciate a toggle for this.

**Comment by daniellee** [comment]
@tmlippman could you explain why you want a toggle? Thanks.

**Comment by tmlippman** [comment]
I'm using grafana dashboards for live monitoring a piece of process equipment. Often you have more panels in a dashboard than can stay in view at one time. But you still want the out-of view panels to update so that the data is current when you do scroll down an look at it. 

The delay for panels to update when scrolling into view is a few seconds, which is just enough to be a minor nuisance. It's worse on slower machines or networks.

**Comment by daniellee** [comment]
@tmlippman thanks for the explanation. Sounds like you have a use case where fast queries are impossible. Your +1 for a toggle is noted. 

The trade-off we made by not having a toggle is that it is a small inconvenience to some users against keeping complexity down. We could have hundreds of toggles but that is something we want to avoid. If enough people complain or if someone has a killer use case then we will of course add a toggle. 

Just out of curiosity, how are you getting the data from your equipment into Grafana?

**Comment by tmlippman** [comment]
The data is stored in an InfluxDB instance on the same server.

**Comment by psellers89** [comment]
We really need  toggle for this. Thanks

**Comment by bobrik** [comment]
At first I was excited, but now I'm just annoyed. Dashboards feel a lot slower with lazy loading.

Here's a gif running in real time speed:

slow

Prometheus responds in ~130ms, yet it takes a couple of seconds for graphs to appear. With lazy loading I'm forced into waiting longer, while without lazy loading I don't see any delay: all graphs appear at the same time. It'd be great if Grafana rendered graphs faster at all times too (is there a ticket already? should I make one?).

Sometimes lazy loading just doesn't trigger from scroll, I have to lift my finger from touchpad / scroll more / look away to make it work. Maybe I'm impatient and just need to wait for a couple of seconds, but life is too short.

**Comment by bobrik** [comment]
Query started at 1758ms, finished around 2192ms, graph appeared after 4583ms:

image

**Comment by bobrik** [comment]
I found the timer that fires 2 seconds after request completion and this looks suspicious:

image

Sadly, I don't know javascript enough to dig deeper into this.

**Comment by daniellee** [comment]
@bobrik that doesn't look right. It is nearly instantaneous for me when scrolling down in a large Graphite dashboard. But I am getting similar behavior to you in this [Prometheus dashboard](http://play.grafana.org/dashboard/db/lazy-loading?refresh=5m&orgId=1) on our play site in Chrome. In Firefox it is as fast as it should be.

Thanks for reporting this. We will investigate and fix.

**Comment by bobrik** [comment]
@daniellee is there an issue I should subscribe to?

**Comment by torkelo** [comment]
we have been troubleshooting this for some time, very strange bug. Seems like Chrome sometimes completely blocks javascript execution when scrolling, happens randomly. Something is blocking (even when disabling all scroll event handlers) this happens randomly 1/3 of the time, especially when you start scrolling during dashboard load all javascript execution is blocked until you stop scrolling and wait for 1-2 seconds. 

Could be something with graph rendering that is blocking while scrolling with Chrome's smooth scrolling enabled. Does not happen in Firefox so seems like Chrome issue.

**Comment by torkelo** [comment]
Openen a new issue to track the progress on this scrolling & blocking behavior. It is not strictly related to lazy loading as you still get the same locking behavior in a large dashboard if you start scrolling during load. 

https://github.com/grafana/grafana/issues/8494

**Comment by torkelo** [comment]
had to revert this due to internal Chrome deadlock bug with scrolling, https://github.com/grafana/grafana/issues/8500

**Pull request merged**